src/NuGet.Core/NuGet.Protocol.Core.v3/Utility/GetDownloadResultUtility.cs

```diff
@@ -86,5 +105,72 @@ public static class GetDownloadResultUtility
 
             throw new InvalidOperationException("Reached an unexpected point in the code");
         }
+
+        private static async Task<DownloadResourceResult> AddPackageDirectAsync(PackageIdentity packageIdentity,
+            Stream packageStream,
+            ISettings settings,
+            string packagesDirectory,
+            ILogger logger,
+            CancellationToken token)
+        {
+            if (packageIdentity == null)
+            {
+                throw new ArgumentNullException(nameof(packageIdentity));
+            }
+
+            if (packageStream == null)
+            {
+                throw new ArgumentNullException(nameof(packageStream));
+            }
+
+            if (settings == null)
+            {
+                throw new ArgumentNullException(nameof(settings));
+            }
+
+            var versionFolderPathContext = new VersionFolderPathContext(
```

Comment:
This approach can generate a file name that exceeds MaxPathLen for very long package names.
  e.g. %userprofile%\AppData\Local\Temp\NuGet\TempCache\<guid>\id\version\id.version.nupkg
We only need a place to temporarily download the nupkg to, so any file name will do.